code/modules/mob/language/language.dm

```diff
@@ -41,8 +42,38 @@
 /datum/language
 	var/list/scramble_cache = list()
 
-/datum/language/proc/scramble(var/input)
+/datum/language/proc/scramble(var/input, var/list/known_languages)
+	var/understand_chance = 0
+	for(var/datum/language/L in known_languages)
+		if(partial_understanding && partial_understanding[L.name])
+			understand_chance += partial_understanding[L.name]
+		if(L.partial_understanding && L.partial_understanding[name])
+			understand_chance += L.partial_understanding[name] * 0.5
+	var/scrambled_text = ""
+	var/list/words = splittext(input, " ")
+	for(var/w in words)
+		if(prob(understand_chance))
+			scrambled_text += " [w] "
+		else
+			var/nword = scramble_word(w)
+			var/ending = copytext(scrambled_text, length(scrambled_text)-1)
+			if(findtext(ending,"."))
```

Comment:
Should this check for other full-stops, like "?" and "!" ?

Reply:
It probably should, yes.